### ISSUE TITLE
How to set this up correctly?

### ISSUE THREAD

**Issue by Rajesh-Yogeshwar** [maciej-gol/tenant-schemas-celery#15]
I intend to use `django_celery_beat` in that case how to configure system to use `DatabaseScheduler` provided by `django_celery_beat`.

After following the example found i tried to run an async task by `apply_async` method but the task could not complete as the task was looking for `Tenant` model in current(tenant) schema whereas it is a part of `public` schema. 

I am under belief that if I use DatabaseScheduler, I have to instances of scheduled task per tenant in their respective schemas instead of public schema. How do I achieve that?

When I try to use it, it breaks on line **50** in `switch_schema` function in app.py. This line `tenant = get_tenant_model().objects.get(schema_name=schema)`

![screenshot from 2017-07-29 15-41-17](https://user-images.githubusercontent.com/7311146/28743981-94e48318-7474-11e7-83bf-8f6edb89ba46.png)

**Comment by maciej-gol** [comment]
Did you add your `Tenant` model to shared apps?

**Comment by Rajesh-Yogeshwar** [comment]
Yes. I have been using this setup for past 6 months. Its working fine. I just had a need to add celery and got it stuck.

**Comment by Rajesh-Yogeshwar** [comment]
@maciej-gol Any update or idea what might be the issue. I never mentioned but I am using celery v4.0.2

**Comment by maciej-gol** [comment]
Sorry, will look into this perhaps this evening. Had a lot of stuff
happening lately.

2017-08-06 7:59 GMT+02:00 Rajesh-Yogeshwar <notifications@github.com>:

**Comment by Rajesh-Yogeshwar** [comment]
Alright. Even I am still probing for the cause. In case if I find it I will update.

**Comment by maciej-gol** [comment]
I couldn't reproduce your issue. Did you migrate the public schema with `python manage.py migrate_schemas --shared`?

**Comment by Rajesh-Yogeshwar** [comment]
Yes I have been running this setup for past six months. Can you please help me understand, how will celery work if I use this module? I mean an example of asynchronous task, a scheduled task. I have been unable to find any material on this one. I have set it up exactly as you have put on README. But it does not work for me, sadly.

**Comment by maciej-gol** [comment]
Basically, whenever a task is about to run the schema is changed to `public`, then a proper schema is taken from the task's metadata and current schema is changed to that. After the task has finished, the schema is reverted to the one before task was run - usually `public`. You can see it here: https://github.com/maciej-gol/tenant-schemas-celery/blob/b85014b5ad4eba44f846913c2d7154efaab71c17/tenant_schemas_celery/app.py#L71 .

I've updated the test app to include an example periodic task. You need to have the virtualenv set up. Then you just sync shared tenants:

```bash
$ python manage.py migrate_schemas --shared
```

Create a tenant:
```bash
$ python manage.py shell
>>> from test_app.shared.models import Client
>>> Client.objects.create(name='test', domain_url='test', schema_name='test')
```

Install `django-celery-beat` and add it to `INSTALLED_APPS` and `SHARED_APPS`.
```bash
$ pip install django-celery-beat
```

Run celery beat:
```bash
$ celery -A celery_app:app beat  --loglevel=INFO -S django
```

Run a celery worker:
```bash
$ celery worker -A celery_app:app -l INFO
```

You should be getting the task's schema printed every 5 seconds or so in the window with the celery worker:

```
[2017-08-08 20:10:04,049: INFO/MainProcess] Received task: test_app.tenant.tasks.periodic_print_schema[faa0162a-10cc-45f9-902b-f338512f7774]  
[2017-08-08 20:10:04,065: WARNING/ForkPoolWorker-2] public
[2017-08-08 20:10:04,065: INFO/ForkPoolWorker-2] Task test_app.tenant.tasks.periodic_print_schema[faa0162a-10cc-45f9-902b-f338512f7774] succeeded in 0.014998486s: None
[2017-08-08 20:10:09,020: INFO/MainProcess] Received task: test_app.tenant.tasks.periodic_print_schema[099c75cf-5d83-4eed-8015-02ad5a344aa6]  
[2017-08-08 20:10:09,060: WARNING/ForkPoolWorker-4] public
[2017-08-08 20:10:09,060: INFO/ForkPoolWorker-4] Task test_app.tenant.tasks.periodic_print_schema[099c75cf-5d83-4eed-8015-02ad5a344aa6] succeeded in 0.034156564001s: None
```

**Comment by Rajesh-Yogeshwar** [comment]
Although it still did not work for me. The output that you have shared, it only prints **public** as schema name. Shouldn't it print **test**?
Secondly, the issue I am facing is that whenever I use **apply_async** or **delay** the code where you switch schemas is giving problems because it is trying to lookup for my tenant model in tenant schema rather than public schema.

`SELECT "customer_customtenantmixin"."id", "customer_customtenantmixin"."domain_url", "customer_customtenantmixin"."schema_name", "customer_tenant"."customtenantmixin_ptr_id", "customer_tenant"."name", "customer_tenant"."slug", "customer_tenant"."subscribed_on", "customer_tenant"."subscription_mode", "customer_tenant"."subscription_ends_on", "customer_tenant"."logo" FROM "customer_tenant" INNER JOIN "customer_customtenantmixin" ON ("customer_tenant"."customtenantmixin_ptr_id" = "customer_customtenantmixin"."id") WHERE "customer_customtenantmixin"."schema_name" = 'prison'`

Above is the query that gets executed, i can see it on console.

**Comment by maciej-gol** [comment]
It prints `public` because the periodic tasks do not have a schema defined, thus it defaults to `public`. When calling the `print_schema` from within a tenant context a proper schema name is printed:

Creating a `Client`:
```python
>>> Client.objects.create(name='test', schema_name='test', domain_url='test')
>>> for client in Client.objects.all():
>>>      with tenant_context(client):
>>>          print_schema.delay()
```

Output in the celery worker:
```
[2017-08-09 06:11:58,095: INFO/MainProcess] Received task: test_app.tenant.tasks.print_schema[751196e0-14de-433f-aa00-bc33f9757404]  
[2017-08-09 06:11:58,114: WARNING/ForkPoolWorker-2] test
[2017-08-09 06:11:58,114: INFO/ForkPoolWorker-2] Task test_app.tenant.tasks.print_schema[751196e0-14de-433f-aa00-bc33f9757404] succeeded in 0.0175698950006s: None
```

I've tested it on the newest `django-tenant-schemas==1.9.0`.

**Comment by Rajesh-Yogeshwar** [comment]
I have set this up on docker, with redis as broker and yes I have just ran it on a fresh DB.  This is relevant part of  my settings file.
![screenshot from 2017-08-09 23-28-39](https://user-images.githubusercontent.com/7311146/29136341-9169eedc-7d5a-11e7-8c41-b3634b79a113.png)

Celery app is set up as per your instructions. In my view, I tried calling it every possible way, but to no avail.

**Comment by maciej-gol** [comment]
Just one more thing - are you using the `django-tenant-schemas` package, or the `django-tenants` package?

**Comment by Rajesh-Yogeshwar** [comment]
I am using `django-tenant-schemas`. Did you find something odd with my settings? In case these are the other necessary variables for you to look through.

```
MULTI_TENANCY_RELATED_APPS = [
	'ppp.customer.apps.CustomerConfig'
]

DJANGO_APPS = [
	
	'django.contrib.auth',
	'django.contrib.admin',
	'django.contrib.sites',
	'django.contrib.humanize',
	'django.contrib.messages',
	'django.contrib.sessions',
	'django.contrib.staticfiles',	
	'django.contrib.contenttypes'
]


THIRD_PARTY_APPS = [
	'mptt',
	'channels',
	'tenant_schemas_celery'
]
```

**Comment by Rajesh-Yogeshwar** [comment]
@maciej-gol Apologies if I am asking too much, can you please try to set it up with docker containers?

**Comment by maciej-gol** [comment]
Do you have `SHARED_APPS` and `TENANT_APPS` settings set? It's described in the docs: https://django-tenant-schemas.readthedocs.io/en/latest/install.html#configure-tenant-and-shared-applications .

**Comment by Rajesh-Yogeshwar** [comment]
Yes, here they are.
```
SHARED_APPS = tuple(['tenant_schemas', 'myapp.customer', 'django_celery_beat', 'myapp.taskapp'] + DJANGO_APPS)
TENANT_APPS = tuple(DJANGO_APPS + THIRD_PARTY_APPS + LOCAL_APPS)
```

**Comment by maciej-gol** [comment]
I'll try with the containers setup, but i'm pretty busy until the end of the week, unfortunately.

**Comment by Rajesh-Yogeshwar** [comment]
Its perfectly alright.

**Comment by Rajesh-Yogeshwar** [comment]
@maciej-gol any update?

**Comment by maciej-gol** [comment]
Yes, sorry. Im unrusting my docker skills :D

23 sie 2017 14:35 "Rajesh-Yogeshwar" <notifications@github.com> napisał(a):

**Comment by Rajesh-Yogeshwar** [comment]
:laughing: I will be patient then.

**Comment by Rajesh-Yogeshwar** [comment]
My guess is that whenever I use `delay/apply_async` the tasks are queue in Redis which is another container. Now when the time comes to execute them, the connection to postgres container cannot be established and hence it erupts with relation could not be found. What do you think? Can it be a reason? Because synchronous execution works fine. If yes, then I think I might need to ask this question somewhere else.

**Comment by Rajesh-Yogeshwar** [comment]
Happy to say that my guess was right. Never been so happy. So the conclusion is right. For the whole delayed task to be executed it was necessary for my redis container to point to my postgres container. I did some experimentation and voila its working now. 

A HUGE THANKS for being patient and putting in your efforts to help me out. Relieved.

**Comment by maciej-gol** [comment]
So you are saying it was a faulty connection to postgres from redis/celery
worker? Perhaps a different user, thus it didn't have the relation set up?
If it couldn't connect at all, you would end up with a DatabaseError.

2017-08-24 10:36 GMT+02:00 Rajesh-Yogeshwar <notifications@github.com>:

**Comment by Rajesh-Yogeshwar** [comment]
I would not say faulty. But I usually setup environment variables for my DB connections and other relevant services and make this variables available to all the services that need them. I was stupid enough not to make this environment variable available to my `redis` service(container). Probably, when the task was to be executed it must have been looking up for database parameters (connection string, username) which were not available as I had not made them available. Once I made those available its started working. I do not know why I did not think this before.
